File: airbyte-commons-temporal-core/src/test/java/io/airbyte/commons/temporal/WorkflowClientWrappedTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.temporal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.airbyte.metrics.lib.MetricClient;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import io.temporal.api.workflowservice.v1.DescribeWorkflowExecutionRequest;
import io.temporal.api.workflowservice.v1.DescribeWorkflowExecutionResponse;
import io.temporal.api.workflowservice.v1.WorkflowServiceGrpc.WorkflowServiceBlockingStub;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.client.WorkflowStub;
import io.temporal.serviceclient.WorkflowServiceStubs;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class WorkflowClientWrappedTest {

  static class MyWorkflow {}

  private static final int maxAttempt = 3;
  private static final int backoffDelayInMillis = 1;
  private static final int backoffMaxDelayInMillis = 10;
  private MetricClient metricClient;
  private WorkflowServiceStubs temporalWorkflowServiceStubs;
  private WorkflowServiceBlockingStub temporalWorkflowServiceBlockingStub;
  private WorkflowClient temporalWorkflowClient;
  private WorkflowClientWrapped workflowClient;

  @BeforeEach
  void beforeEach() {
    metricClient = mock(MetricClient.class);
    temporalWorkflowServiceBlockingStub = mock(WorkflowServiceBlockingStub.class);
    temporalWorkflowServiceStubs = mock(WorkflowServiceStubs.class);
    when(temporalWorkflowServiceStubs.blockingStub()).thenReturn(temporalWorkflowServiceBlockingStub);
    temporalWorkflowClient = mock(WorkflowClient.class);
    when(temporalWorkflowClient.getWorkflowServiceStubs()).thenReturn(temporalWorkflowServiceStubs);
    workflowClient = new WorkflowClientWrapped(temporalWorkflowClient, metricClient, maxAttempt, backoffDelayInMillis, backoffMaxDelayInMillis);
  }

  @Test
  void testRetryLogic() {
    when(temporalWorkflowClient.newWorkflowStub(any(), anyString()))
        .thenThrow(unavailable());

    assertThrows(StatusRuntimeException.class, () -> workflowClient.newWorkflowStub(MyWorkflow.class, "fail"));
    verify(temporalWorkflowClient, times(3)).newWorkflowStub(any(), anyString());
    verify(metricClient, times(2)).count(any(), anyLong(), any(), any(), any());
  }

  @Test
  void testNewWorkflowStub() {
    final MyWorkflow expected = new MyWorkflow();
    when(temporalWorkflowClient.newWorkflowStub(any(), anyString()))
        .thenThrow(unavailable())
        .thenReturn(expected);

    final MyWorkflow actual = workflowClient.newWorkflowStub(MyWorkflow.class, "woot");
    assertEquals(expected, actual);
  }

  @Test
  void testNewWorkflowStubWithOptions() {
    final MyWorkflow expected = new MyWorkflow();
    when(temporalWorkflowClient.newWorkflowStub(any(), (WorkflowOptions) any()))
        .thenThrow(unavailable())
        .thenReturn(expected);

    final MyWorkflow actual = workflowClient.newWorkflowStub(MyWorkflow.class, WorkflowOptions.getDefaultInstance());
    assertEquals(expected, actual);
  }

  @Test
  void testTerminateWorkflow() {
    final var workflowStub = mock(WorkflowStub.class);
    when(temporalWorkflowClient.newUntypedWorkflowStub(anyString()))
        .thenThrow(unavailable())
        .thenReturn(workflowStub);

    workflowClient.terminateWorkflow("workflow", "test terminate");
    verify(temporalWorkflowClient, times(2)).newUntypedWorkflowStub("workflow");
    verify(workflowStub).terminate("test terminate");
  }

  @Test
  void testBlockingDescribeWorkflowExecution() {
    final DescribeWorkflowExecutionResponse expected = mock(DescribeWorkflowExecutionResponse.class);
    when(temporalWorkflowServiceBlockingStub.describeWorkflowExecution(any()))
        .thenThrow(unavailable())
        .thenReturn(expected);

    final DescribeWorkflowExecutionResponse actual = workflowClient.blockingDescribeWorkflowExecution(mock(DescribeWorkflowExecutionRequest.class));
    assertEquals(expected, actual);
  }

  @Test
  void testSignalsAreNotRetried() {
    when(temporalWorkflowClient.signalWithStart(any())).thenThrow(unavailable());
    assertThrows(StatusRuntimeException.class, () -> {
      final var request = workflowClient.newSignalWithStartRequest();
      workflowClient.signalWithStart(request);
    });
    verify(temporalWorkflowClient, times(1)).signalWithStart(any());
  }

  private static StatusRuntimeException unavailable() {
    return new StatusRuntimeException(Status.UNAVAILABLE);
  }

}

File: airbyte-commons-temporal-core/src/test/kotlin/io/airbyte/commons/temporal/WorkflowClientWrappedTest.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.temporal

import io.airbyte.metrics.MetricClient
import io.grpc.Status
import io.grpc.StatusRuntimeException
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.temporal.api.workflowservice.v1.DescribeWorkflowExecutionRequest
import io.temporal.api.workflowservice.v1.DescribeWorkflowExecutionResponse
import io.temporal.api.workflowservice.v1.WorkflowServiceGrpc.WorkflowServiceBlockingStub
import io.temporal.client.WorkflowClient
import io.temporal.client.WorkflowOptions
import io.temporal.client.WorkflowStub
import io.temporal.serviceclient.WorkflowServiceStubs
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class WorkflowClientWrappedTest {
  internal class MyWorkflow

  private lateinit var metricClient: MetricClient
  private lateinit var temporalWorkflowServiceStubs: WorkflowServiceStubs
  private lateinit var temporalWorkflowServiceBlockingStub: WorkflowServiceBlockingStub
  private lateinit var temporalWorkflowClient: WorkflowClient
  private lateinit var workflowClient: WorkflowClientWrapped

  @BeforeEach
  fun beforeEach() {
    metricClient = mockk(relaxed = true)
    temporalWorkflowServiceBlockingStub = mockk()
    temporalWorkflowServiceStubs = mockk()
    every { temporalWorkflowServiceStubs.blockingStub() } returns temporalWorkflowServiceBlockingStub
    temporalWorkflowClient = mockk()
    every { temporalWorkflowClient.workflowServiceStubs } returns temporalWorkflowServiceStubs
    workflowClient =
      WorkflowClientWrapped(temporalWorkflowClient, metricClient, MAX_ATTEMPT, BACKOFF_DELAY_IN_MILLIS, BACKOFF_MAX_DELAY_IN_MILLIS)
  }

  @Test
  fun testRetryLogic() {
    every { temporalWorkflowClient.newWorkflowStub(any<Class<Any>>(), any<String>()) } throws unavailable()

    Assertions.assertThrows(
      StatusRuntimeException::class.java,
    ) { workflowClient.newWorkflowStub(MyWorkflow::class.java, "fail") }
    verify(exactly = 3) { temporalWorkflowClient.newWorkflowStub(any<Class<Any>>(), any<String>()) }
    verify(exactly = 2) { metricClient.count(any(), any(), *anyVararg()) }
  }

  @Test
  fun testNewWorkflowStub() {
    val expected = MyWorkflow()
    every { temporalWorkflowClient.newWorkflowStub(any<Class<Any>>(), any<String>()) } throws unavailable() andThen expected

    val actual = workflowClient.newWorkflowStub(MyWorkflow::class.java, "woot")
    Assertions.assertEquals(expected, actual)
  }

  @Test
  fun testNewWorkflowStubWithOptions() {
    val expected = MyWorkflow()
    every { temporalWorkflowClient.newWorkflowStub(any<Class<Any>>(), any<WorkflowOptions>()) } throws unavailable() andThen expected

    val actual = workflowClient.newWorkflowStub(MyWorkflow::class.java, WorkflowOptions.getDefaultInstance())
    Assertions.assertEquals(expected, actual)
  }

  @Test
  fun testTerminateWorkflow() {
    val workflowStub = mockk<WorkflowStub>(relaxed = true)
    every { temporalWorkflowClient.newUntypedWorkflowStub(any<String>()) } throws unavailable() andThen workflowStub

    workflowClient.terminateWorkflow("workflow", "test terminate")
    verify(exactly = 2) { temporalWorkflowClient.newUntypedWorkflowStub("workflow") }
    verify { workflowStub.terminate("test terminate") }
  }

  @Test
  fun testBlockingDescribeWorkflowExecution() {
    val expected = mockk<DescribeWorkflowExecutionResponse>()
    every { temporalWorkflowServiceBlockingStub.describeWorkflowExecution(any()) } throws unavailable() andThen expected

    val actual =
      workflowClient.blockingDescribeWorkflowExecution(
        mockk<DescribeWorkflowExecutionRequest>(),
      )
    Assertions.assertEquals(expected, actual)
  }

  @Test
  fun testSignalsAreNotRetried() {
    every { temporalWorkflowClient.signalWithStart(any()) } throws unavailable()
    every { temporalWorkflowClient.newSignalWithStartRequest() } returns mockk()
    Assertions.assertThrows(StatusRuntimeException::class.java) {
      val request = workflowClient.newSignalWithStartRequest()
      workflowClient.signalWithStart(request)
    }
    verify(exactly = 1) { temporalWorkflowClient.signalWithStart(any()) }
  }

  companion object {
    private const val MAX_ATTEMPT = 3
    private const val BACKOFF_DELAY_IN_MILLIS = 1
    private const val BACKOFF_MAX_DELAY_IN_MILLIS = 10

    private fun unavailable(): StatusRuntimeException = StatusRuntimeException(Status.UNAVAILABLE)
  }
}
